pkg/s3-proxy/server/tls.go

```diff
@@ -0,0 +1,228 @@
+package server
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/oxyno-zeta/s3-proxy/pkg/s3-proxy/config"
+	"github.com/oxyno-zeta/s3-proxy/pkg/s3-proxy/log"
+	"github.com/oxyno-zeta/s3-proxy/pkg/s3-proxy/utils"
+)
+
+// The intersection of the recommended cipher suites from https://ciphersuite.info/cs/?security=recommended
+// and the suites implemented in Go.
+var defaultCipherSuites = []uint16{
+	tls.TLS_AES_128_GCM_SHA256,
+	tls.TLS_AES_256_GCM_SHA384,
+	tls.TLS_CHACHA20_POLY1305_SHA256,
+	tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+	tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+	tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
+}
+
+// The number of bits to allow in a generated certificate serial number.
+const certSerialBits = 128
+
+// The number of bits to use in a generated RSA private key.
+const rsaKeySize = 2048
+
+// generateTLSConfig creates a crypto/tls.Config configuration for a net/http.Server from an s3-proxy
+// ServerSSLConfig.
+func generateTLSConfig(sslConfig *config.ServerSSLConfig, logger log.Logger) (*tls.Config, error) {
+	result := tls.Config{
+		MinVersion:   tls.VersionTLS12,
+		CipherSuites: defaultCipherSuites,
+	}
+
+	if sslConfig == nil {
+		return nil, nil //nolint:nilnil // We do not want a TLS config in this case
+	}
+
+	if sslConfig.Enabled == nil {
+		// Don't enable SSL if no certificates were set or requested.
+		if len(sslConfig.Certificates) == 0 && len(sslConfig.SelfSignedHostnames) == 0 {
+			return nil, nil //nolint:nilnil // We do not want a TLS config in this case
+		}
+	} else if !*sslConfig.Enabled {
+		return nil, nil //nolint:nilnil // We do not want a TLS config in this case
+	}
+
+	if len(sslConfig.SelfSignedHostnames) == 0 && len(sslConfig.Certificates) == 0 {
+		return nil, errors.New("at least one certificate must be specified")
+	}
+
+	// Generate self-signed certificates for each hostname requested
+	if len(sslConfig.SelfSignedHostnames) > 0 {
+		selfSignedCert, err := generateSelfSignedCertificate(sslConfig.SelfSignedHostnames)
+		if err != nil {
+			logger.Errorf("Failed to generate self-signed certificate: %v", err)
+
+			return nil, err
+		}
+
+		result.Certificates = append(result.Certificates, selfSignedCert)
+	}
+
+	// Set min and max TLS versions if they were specified in the config.
+	if sslConfig.MinTLSVersion != nil {
+		result.MinVersion = utils.ParseTLSVersion(*sslConfig.MinTLSVersion)
+		if result.MinVersion == 0 {
+			logger.Fatalf("Invalid TLS version: %v", *sslConfig.MinTLSVersion)
+		}
+	}
+
+	if sslConfig.MaxTLSVersion != nil {
+		result.MaxVersion = utils.ParseTLSVersion(*sslConfig.MaxTLSVersion)
+		if result.MaxVersion == 0 {
+			logger.Fatalf("Invalid TLS version: %v", *sslConfig.MaxTLSVersion)
```

Comment:
That is covered by the configuration validation no? Can it be removed or transformed into a return error?